0012-BIRS-distributed-scheduler.md

```diff
@@ -0,0 +1,611 @@
+![Scheduler APIs (SideCar Facing)](./resources/0004-BIRS-distributed-scheduler/sidecarToSchedulerComm.png)
+
+We will use the same exact protos from the Public Dapr API, but inside a **new**: `dapr/proto/scheduler/scheduler.proto`.
+The Schedule/Get/Delete job(s) will be performed via a unary call to the Scheduler Service.
+
+There is a bidirectional streaming connection between the daprd sidecar and the Scheduler Service to allow for the acknowledgment of successfully triggered jobs.
+
+###### Scheduler Service APIs
+
+In the **new** `dapr/proto/scheduler/scheduler.proto`, the daprd sidecar upon startup will establish a streaming connection with the Scheduler Service such that at the trigger time for a job the Scheduler Service will send that job to the daprd sidecar which is watching for jobs. Then the daprd sidecar will send the job to the app sending the `WatchJobsResponse` back to the Scheduler.
+
+![WatchJobsFLow](./resources/0004-BIRS-distributed-scheduler/watchJobsFlow.png)
+
+```proto
+service Scheduler {
+	// ScheduleJob is used by the daprd sidecar to schedule a job.
+	rpc ScheduleJob(ScheduleJobRequest) returns (ScheduleJobResponse) {}
+	
+	// Get a job
+	rpc GetJob(GetJobRequest) returns (GetJobResponse) {}
+	
+	// DeleteJob is used by the daprd sidecar to delete a job.
+	rpc DeleteJob(DeleteJobRequest) returns (DeleteJobResponse) {}
+	
+	// WatchJobs is used by the daprd sidecar to connect to the Scheduler
+	// service to watch for jobs triggering back.
+	rpc WatchJobs(stream WatchJobsRequest) returns (stream WatchJobsResponse) {}
+}
+
+
+message Job {
+  // The schedule for the job.
+  optional string schedule = 1;
+
+  // Optional: jobs with fixed repeat counts (accounting for Actor Reminders).
+  optional uint32 repeats = 2;
+
+  // Optional: sets time at which or time interval before the callback is invoked for the first time.
+  optional string due_time = 3;
+
+  // Optional: Time To Live to allow for auto deletes (accounting for Actor Reminders).
+  optional string ttl = 4;
+
+  // Job data.
+  google.protobuf.Any data = 5;
+}
+
+// TypeJob is the message used by the daprd sidecar to schedule a job
+// from an App.
+message TypeJob {}
+
+// TypeActorReminder is the message used by the daprd sidecar to
+// schedule a job from an Actor Reminder.
+message TypeActorReminder {
+  // id is the actor ID.
+	string id = 1;
+
+  // type is the actor type.
+	string type = 2;
+}
+
+// JobMetadataType holds the typed metadata associated with the job for
+// different origins.
+message JobMetadataType {
+  oneof type {
+    TypeJob job = 1;
+    TypeActorReminder actor = 2;
+  }
+}
+
+// JobMetadata is the message used by the daprd sidecar to schedule a
+// job.
+message JobMetadata {
+  // app_id is the App ID of the requester.
+	string app_id = 1;
+
+  // namespace is the namespace of the requester.
+  string namespace = 2;
+
+  // type is the type of the job.
+  JobMetadataType type = 3;
```

Comment:
Some better names (IMO):
JobMetadataDetails details;
JobMetadataExtension extension;
JobMetadataInfo info;

Reply:
I will change it to `JobMetadataKind`

Reply:
That too Kubernetes-like.

Reply:
@yaron2 thoughts?

Reply:
Sounds better IMO.

Reply:
I support JobTargetMetadata